### PR TITLE
Feature/add publishing to elasticsearch support

### PR DESCRIPTION
**Pull request by SCBbestof**

### CODE REVIEW

File: vars/publishToElasticsearch.groovy [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+  Environment targetEnv = computeTargetEnvironment(environmentInput)
+  DeploymentUtils deployUtil = new DeploymentUtils()
+
+  String UUIDS_FILE_PATH = "uuids.txt"
+  String INDEX_ZAPPER_APP_NAME = "elasticsearch-index-zapper"
+  String GET_MONGO_CONTAINER_CMD = "kubectl get pods | grep mongodb | awk '{print \$1}' | head -1"
+  String JQ_ALTERNATIVE_CMD = "sed '/uuid/!d' | sed s/\\\"uuid\\\"://g | sed s/\\\"//g | sed s/\\ //g | sed -e 's/[{}]//g' | awk -v RS=':' '{print \$1}'"
+
+  String GET_METHODE_UUIDS_CMD = "mongo upp-store -eval 'rs.slaveOk(); db.content.find({\"mediaType\":null,\"identifiers.authority\": {\$regex: /FTCOM-METHODE/ },\"type\": {\$nin: [\"ContentPackage\",\"Content\",\"ImageSet\"]}}, {_id: false, uuid: 1}).forEach(function(o) { print(o.uuid)})' --quiet > ${UUIDS_FILE_PATH}"
```
Comment (tosan88) [inline vars/publishToElasticsearch.groovy]: Are you sure it will run against a slave MongoDB? If we run against the primary, we might block write access for a while.

File: vars/publishToElasticsearch.groovy [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+private static Environment computeTargetEnvironment(String environmentInput) {
+  Environment targetEnv = EnvsRegistry.getEnvironment(environmentInput)
+  if (targetEnv == null) {
+    throw new IllegalArgumentException("Unknown environment ${environmentInput}. The environment is required.")
+  }
+  return targetEnv
+}
+
+private String getDeliveryClusterUrl(String envName, String region) {
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: Please use Environment.getEntryPointUrl for getting this URL.

File: vars/publishToElasticsearch.groovy [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  String GET_WORDPRESS_UUIDS_CMD = "mongo upp-store -eval 'rs.slaveOk(); db.content.find({\"mediaType\":null,\"identifiers.authority\": {\$regex: /FT-LABS/ }}, {_id: false, uuid: 1}).forEach(function(o) { printjson(o)})' --quiet | ${JQ_ALTERNATIVE_CMD} > ${UUIDS_FILE_PATH}"
+  String GET_VIDEO_UUIDS_CMD = "mongo upp-store -eval 'rs.slaveOk(); db.content.find({\"identifiers.authority\": {\$regex: /NEXT-VIDEO-EDITOR/ }}, {_id: false, uuid: 1}).forEach(function(o) { printjson(o)})' --quiet | ${JQ_ALTERNATIVE_CMD} > ${UUIDS_FILE_PATH}"
+
+  Closure SERVICE_SHUTDOWN = { sh "kubectl scale --replicas=0 deployments/content-rw-elasticsearch" }
+
+  String AWS_ACCESS_KEY
+  String AWS_SECRET_KEY
+  String ES_ENDPOINT
+
+  Closure GET_CONFIGURATION = {
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: Instead of having closures as constants/vars, I think it would be better to have methods that call inside runWith... with the inlined closure. This way we can express better the functionality by using method names.

File: vars/publishToElasticsearch.groovy [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  Environment targetEnv = computeTargetEnvironment(environmentInput)
+  DeploymentUtils deployUtil = new DeploymentUtils()
+
+  String UUIDS_FILE_PATH = "uuids.txt"
+  String INDEX_ZAPPER_APP_NAME = "elasticsearch-index-zapper"
+  String GET_MONGO_CONTAINER_CMD = "kubectl get pods | grep mongodb | awk '{print \$1}' | head -1"
+  String JQ_ALTERNATIVE_CMD = "sed '/uuid/!d' | sed s/\\\"uuid\\\"://g | sed s/\\\"//g | sed s/\\ //g | sed -e 's/[{}]//g' | awk -v RS=':' '{print \$1}'"
+
+  String GET_METHODE_UUIDS_CMD = "mongo upp-store -eval 'rs.slaveOk(); db.content.find({\"mediaType\":null,\"identifiers.authority\": {\$regex: /FTCOM-METHODE/ },\"type\": {\$nin: [\"ContentPackage\",\"Content\",\"ImageSet\"]}}, {_id: false, uuid: 1}).forEach(function(o) { print(o.uuid)})' --quiet > ${UUIDS_FILE_PATH}"
+  String GET_WORDPRESS_UUIDS_CMD = "mongo upp-store -eval 'rs.slaveOk(); db.content.find({\"mediaType\":null,\"identifiers.authority\": {\$regex: /FT-LABS/ }}, {_id: false, uuid: 1}).forEach(function(o) { printjson(o)})' --quiet | ${JQ_ALTERNATIVE_CMD} > ${UUIDS_FILE_PATH}"
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: These mongo one liners are really hard to follow. Is there a way we can make them more intelligible ?

File: vars/publishToElasticsearch.groovy [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    ES_ENDPOINT = sh(
+            script: "kubectl get configmap global-config -o yaml | grep aws.content.elasticsearch.endpoint | awk '{print \$2}'",
+            returnStdout: true
+    ).trim()
+  }
+
+  Closure INDEX_DELETION = {
+    git credentialsId: 'ft-upp-team', url: 'git@github.com:Financial-Times/elasticsearch-index-zapper.git', branch: 'master'
+
+    sh "mkdir -p \$GOPATH/src/${INDEX_ZAPPER_APP_NAME} && mv ./* \$GOPATH/src/${INDEX_ZAPPER_APP_NAME} && " +
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: It would be great to have the Docker image with the executable already prepared.

File: vars/publishToElasticsearch.groovy (hunk 4 above)
Comment (sbuliarca) [reply vars/publishToElasticsearch.groovy]: Now that I'm thinking about it, probably a cleaner solution would be to have a microservice deployed on the cluster that exposes endpoints that give you this information. I know it involves much more work but it might be the proper way to do it.

File: vars/publishToElasticsearch.groovy [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      }
+
+      stage('Enable content-rw-elasticsearch') {
+        deployUtil.runWithK8SCliTools(targetEnv, Cluster.DELIVERY, regionInput, SERVICE_STARTUP)
+      }
+
+      stage('Get UUIDs and publish content') {
+        echo "CMS Input: ${cmsInput}"
+        String[] cmsArray = cmsInput.split(",")
+        echo "CMS Array: ${cmsArray}"
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: Probably too much echo

File: vars/publishToElasticsearch.groovy [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        for (int i = 0; i < cmsArray.length; i++) {
+          echo "CMS: ${cmsArray[i]}"
+          sh "rm -f uuids.txt"
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: The constant can be used here

File: vars/publishToElasticsearch.groovy [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+          echo 'Skipping ES index zapping'
+        }
+      }
+
+      stage('Enable content-rw-elasticsearch') {
+        deployUtil.runWithK8SCliTools(targetEnv, Cluster.DELIVERY, regionInput, SERVICE_STARTUP)
+      }
+
+      stage('Get UUIDs and publish content') {
+        echo "CMS Input: ${cmsInput}"
```
Comment (sbuliarca) [inline vars/publishToElasticsearch.groovy]: Just as a side note: this could also be expressed with multiple boolean input parameters

File: vars/publishToElasticsearch.groovy [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  node('docker') {
+    catchError {
+      stage('Disable content-rw-elasticsearch') {
+        deployUtil.runWithK8SCliTools(targetEnv, Cluster.DELIVERY, regionInput, SERVICE_SHUTDOWN)
+        sleep(5) // Wait 5s for content-rw-elasticsearch to be disabled
+      }
+
+      stage('Delete ES index') {
+        if (zapIndexInput) {
+          echo "INDEX DELETION BEFORE: ${AWS_ACCESS_KEY}"
```
[the remark this answers is not in the document]
Comment (SCBbestof) [reply vars/publishToElasticsearch.groovy]: Removed

File: vars/publishToElasticsearch.groovy (hunk 6 above)
Comment (SCBbestof) [reply vars/publishToElasticsearch.groovy]: Removed. That was there for debugging purposes

File: vars/publishToElasticsearch.groovy (hunk 7 above)
Comment (SCBbestof) [reply vars/publishToElasticsearch.groovy]: Done

File: vars/publishToElasticsearch.groovy (hunk 3 above)
Comment (SCBbestof) [reply vars/publishToElasticsearch.groovy]: Done

**Comment by SCBbestof** [comment]
[Automatize SAPI reindexing](https://trello.com/c/RzNGDdpz/125-automatize-sapi-reindexing)
